### PR TITLE
[1.16] Enabling Player based crafting recipes autocrafting

### PR DESCRIPTION
**Pull request by mcenderdragon**
This adds a player to the Pattern Container & stored the creator of a pattern in the NBT of the item.
And it adds more info to the DummyCraftingInventory so recipes which are player based (for example futurepack) can use this player/UUD and research recipes are fianlly available for autocrafting.

I tried to add the player who requested a crafting to everythign but noticed: I only check thath in the match function so when the Pattern is laoded from the item, match is then never checked again so it would help my problem here.

### CODE REVIEW

**Review by Darkere — CHANGES_REQUESTED** [review]
Had a quick look. Haven't tried anything yet though.

File: src/main/java/com/refinedmods/refinedstorage/api/network/INetwork.java [hunk 1]
```diff
@@ -1,5 +1,9 @@
 package com.refinedmods.refinedstorage.api.network;
 
+import java.util.function.Predicate;
+
+import javax.annotation.Nonnull;
+
```
[the remark this answers is not in the document]
Comment (mcenderdragon) [reply src/main/java/com/refinedmods/refinedstorage/api/network/INetwork.java:6]: fixed

File: build.gradle [hunk 2]
```diff
@@ -25,6 +25,7 @@ repositories {
     maven {
         url = "https://maven.theillusivec4.top/"
     }
+
```
[the remark this answers is not in the document]
Comment (mcenderdragon) [reply build.gradle:28]: fixed

File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/node/GridNetworkNode.java [hunk 3]
```diff
@@ -383,6 +398,7 @@ public FluidInventory getProcessingMatrixFluids() {
 
     @Override
     public void onCraftingMatrixChanged() {
+
```
[the remark this answers is not in the document]
Comment (mcenderdragon) [reply src/main/java/com/refinedmods/refinedstorage/apiimpl/network/node/GridNetworkNode.java]: fixed

**Comment by mcenderdragon** [comment]
to test add 
```
 maven {
        url "https://www.moadmaven.dev.com"
        content {
            includeGroup "curse.maven"
        }
    }
```
And 

```
	runtimeOnly fg.deobf("redsnakegames.mcenderdragon:futurepack:1.16.4-31.2.6878")
	compileOnly fg.deobf("redsnakegames.mcenderdragon:futurepack:1.16.4-31.2.6878:api") 
```

To the build script. ANd then its posssible to craft patterns with futurepack recipes & also craft them. use `/futurepack research Dev unlock all` to get all researches so the recipes are available to you.

**Comment by Darkere** [comment]
So uh... I took a closer look at the code. This is... a hack to get support for your own mod in. 

At no point does your code interact with the underlying forge/minecraft code to make things compatible with other mods using the crafting system. Compatibility with other mods is something we want, but not by having random unused fields in classes. 

Sorry I don't think this is going to have a chance like it is.

**Comment by mcenderdragon** [comment]
> 
> 
> So uh... I took a closer look at the code. This is... a hack to get support for your own mod in.
> 
> At no point does your code interact with the underlying forge/minecraft code to make things compatible with other mods using the crafting system. Compatibility with other mods is something we want, but not by having random unused fields in classes.
> 
> Sorry I don't think this is going to have a chance like it is.

That is true. I am open to other implementations, the problem is Forge has a "current crafting player" in ForgeHooks, but the time this is used is to late since there is no player when the patterns are loaded and the "matches" function is called in IRecipe. I added those fields to pass more information to the IRecipe since that only gets the CraftingInventory

**Comment by mcenderdragon** [comment]
Current issue with latest changes: it doenst work client side, because of the pattern cahce and the fake player is only set on server side.

**Review by raoulvdberge — CHANGES_REQUESTED** [review]
Can this be re-targeted for 1.18 (`develop`)

**Comment by raoulvdberge** [comment]
This is still targeting 1.16. Feel free to open a new PR for `develop`

**Pull request closed without merging**